File: LogiOS/TrustarApp/Database/user/Entity/CollectionResult.swift
//
//  CollectionResult.swift
//  TrustarApp
//
//  Created by DionSoftware on 2022/01/19.
//

import Foundation
import GRDB

enum CollectionResultColumns: String, ColumnExpression {
    case allocationNo, allocationRowNo, json
    case collectionNo, collectionCd, collectionNm, expectedQuantity, actualQuantity, collectionClass, displayOrder
}

class CollectionResult: BaseEntity, Codable {
    var allocationNo: String = Resources.strEmpty
    var allocationRowNo: Int = Resources.zeroNumber
    var json: [RowCollectionResult] = []
    
    override init() {
        super.init()
    }
    
    init(allocationNo: String,
         allocationRowNo: Int,
         rows: [RowCollectionResult]
    ) {
        self.allocationNo = allocationNo
        self.allocationRowNo = allocationRowNo
        self.json = rows
        super.init()
    }
    
    override class var databaseTableName: String {"CollectionResult"}
    override class var primaryKey: String {"allocationNo,allocationRowNo"}
    
    required init(row: Row) {
        allocationNo = row[CollectionResultColumns.allocationNo]
        allocationRowNo = row[CollectionResultColumns.allocationRowNo]
        json = Helper.Shared.convertJsonToObject(json: row[CollectionResultColumns.json])
        super.init(row: row)
    }
    
    override func encode(to container: inout PersistenceContainer) {
        container[CollectionResultColumns.allocationNo] = allocationNo
        container[CollectionResultColumns.allocationRowNo] = allocationRowNo
        container[CollectionResultColumns.json] = Helper.Shared.convertObjectToJson(body: json)
        super.encode(to: &container)
    }
    
    func rows() -> [RowCollectionResult] {
        var list: [RowCollectionResult] = []
        for j in json {
            list.append(RowCollectionResult(collectionNo: j.collectionNo,
                                            collectionCd: j.collectionCd,
                                            collectionNm: j.collectionNm,
                                            expectedQuantity: j.expectedQuantity,
                                            actualQuantity: j.actualQuantity,
                                            collectionClass: j.collectionClass,
                                            displayOrder: j.displayOrder))
        }
        
        return list
    }
    
    class RowCollectionResult: Codable {
        var collectionNo: String?
        var collectionCd: String?
        var collectionNm: String?
        var expectedQuantity: Double
        var actualQuantity: Double
        var collectionClass: Int
        var displayOrder: Int
        
        init(collectionNo: String?,
             collectionCd: String?,
             collectionNm: String?,
             expectedQuantity: Double,
             actualQuantity: Double,
             collectionClass: Int,
             displayOrder: Int
        ){
            self.collectionNo = collectionNo
            self.collectionCd = collectionCd
            self.collectionNm = collectionNm
            self.expectedQuantity = expectedQuantity
            self.actualQuantity = actualQuantity
            self.collectionClass = collectionClass
            self.displayOrder = displayOrder
        }
        
        func emptyCd() -> Bool {
            return collectionCd == nil || collectionCd!.isEmpty
        }

        func newRow(quantity: Double, name: String) -> RowCollectionResult {
            return RowCollectionResult(collectionNo: collectionNo,
                                       collectionCd: collectionCd,
                                       collectionNm: emptyCd() ? name : collectionNm,
                                       expectedQuantity: expectedQuantity,
                                       actualQuantity: quantity,
                                       collectionClass: collectionClass,
                                       displayOrder: displayOrder)
        }
    }
}

File: LogAndroid/TrustarApp/src/main/kotlin/jp/co/toukei/log/trustar/db/user/entity/CollectionResult.kt
package jp.co.toukei.log.trustar.db.user.entity

import androidx.room.ColumnInfo
import androidx.room.Entity
import jp.co.toukei.log.lib.jsonArrNotNull
import jp.co.toukei.log.lib.room.AbstractSync
import jp.co.toukei.log.trustar.forEachOptObject
import org.json.JSONArray
import third.jsonObj

@Entity(
        tableName = "collection_result",
        primaryKeys = ["allocation_no", "allocation_row_no"]
)
class CollectionResult(
        @ColumnInfo(name = "allocation_no") @JvmField val allocationNo: String,
        @ColumnInfo(name = "allocation_row_no") @JvmField val allocationRowNo: Int,
        @ColumnInfo(name = "json") @JvmField val json: JSONArray, // No pk for rows.
) : AbstractSync() {

    constructor(
            allocationNo: String,
            allocationRowNo: Int,
            rows: List<Row>
    ) : this(allocationNo, allocationRowNo, rows.jsonArrNotNull {
        jsonObj {
            "collectionNo" v it.collectionNo
            "collectionCd" v it.collectionCd
            "collectionNm" v it.collectionNm
            "expectedQuantity" v it.expectedQuantity
            "actualQuantity" v it.actualQuantity
            "collectionClass" v it.collectionClass
            "displayOrder" v it.displayOrder
        }
    })

    fun rows(): List<Row> {
        val list = ArrayList<Row>()
        json.forEachOptObject { _, j ->
            if (j != null)
                list += Row(
                        j.optString("collectionNo"),
                        j.optString("collectionCd"),
                        j.optString("collectionNm"),
                        j.getDouble("expectedQuantity"),
                        j.getDouble("actualQuantity"),
                        j.getInt("collectionClass"),
                        j.getInt("displayOrder"),
                )
        }
        return list
    }

    class Row(
            @JvmField val collectionNo: String?,
            @JvmField val collectionCd: String?,
            @JvmField val collectionNm: String?,
            @JvmField val expectedQuantity: Double,
            @JvmField val actualQuantity: Double,
            @JvmField val collectionClass: Int,
            @JvmField val displayOrder: Int,
    ) {

        fun emptyCd(): Boolean = collectionCd.isNullOrEmpty()

        fun newRow(quantity: Double, name: String) = Row(
                collectionNo,
                collectionCd,
                if (emptyCd()) name else collectionNm,
                expectedQuantity,
                quantity,
                collectionClass,
                displayOrder
        )
    }
}
